File: app/src/main/java/io/github/adulescentia/LUMOS_lib/GestureStateManager.java
package io.github.adulescentia.LUMOS_lib;

import android.util.Log;

public class GestureStateManager {
    private static final String TAG = "GestureStateManager";

    public enum Gesture {
        FIST,
        PALM,
        ONE_FINGER,
        V_SIGN,
        UNDEF
    }

    public interface ActionListener {
        void onDeviceSelectionToggled(boolean isSelected);
        void onDevicePowerToggled();
        void onDeviceModeApplied(float modeValue);
    }

    private Gesture prevGesture = Gesture.UNDEF;
    private boolean isDeviceSelected = false;
    private boolean isTrackingModeActive = false;

    private float baseWristY = 0.0f;
    private float currentModeValue = 0.0f;
    private float sensitivity = 1.0f;

    private ActionListener actionListener;

    public void setActionListener(ActionListener actionListener) {
        this.actionListener = actionListener;
    }

    public void setSensitivity(float sensitivity) {
        this.sensitivity = sensitivity;
    }

    public boolean isDeviceSelected() {
        return isDeviceSelected;
    }

    public boolean isTrackingModeActive() {
        return isTrackingModeActive;
    }

    public void update(Gesture currentGesture, float currentWristY) {
        boolean isComingFromReadyState = (prevGesture == Gesture.FIST || prevGesture == Gesture.UNDEF);

        switch (currentGesture) {
            case ONE_FINGER:
                if (isComingFromReadyState && prevGesture != Gesture.ONE_FINGER) {
                    isDeviceSelected = !isDeviceSelected;
                    if (isDeviceSelected) {
                        Log.d(TAG, "🎯 [EVENT] 디바이스가 선택(조준 완료) 되었습니다.");
                        isTrackingModeActive = false;
                    } else {
                        Log.d(TAG, "❌ [EVENT] 디바이스 선택이 해제되었습니다.");
                    }
                    if (actionListener != null) {
                        actionListener.onDeviceSelectionToggled(isDeviceSelected);
                    }
                }
                break;

            case PALM:
                if (prevGesture == Gesture.FIST) {
                    Log.d(TAG, "🔌 [EVENT] 디바이스 전원(On/Off) 토글 명령이 실행되었습니다.");
                    if (actionListener != null) {
                        actionListener.onDevicePowerToggled();
                    }
                }
                break;

            case V_SIGN:
                if (isComingFromReadyState && prevGesture != Gesture.V_SIGN) {
                    if (isDeviceSelected) {
                        isTrackingModeActive = !isTrackingModeActive;

                        if (isTrackingModeActive) {
                            baseWristY = currentWristY;
                            currentModeValue = 0.0f;
                            Log.d(TAG, "📏 [MODE] 모드 조절 활성화. 기준 Y: " + baseWristY + " (감도: " + sensitivity + ")");
                        } else {
                            Log.d(TAG, "💾 [MODE] 모드 조절 비활성화. 최종 높이 변화 mode 반영: " + currentModeValue);
                            applyDeviceMode(currentModeValue);
                        }
                    } else {
                        Log.w(TAG, "⚠️ [WARN] 선택된 디바이스가 없어 모드 조절을 시작할 수 없습니다.");
                    }
                }
                break;

            case FIST:
            case UNDEF:
                break;
        }

        if (isTrackingModeActive) {
            float deltaY = baseWristY - currentWristY;
            currentModeValue = deltaY * sensitivity;
            Log.d(TAG, "🔄 [TRACKING] 실시간 높이 변화량 추적 중... deltaY(순수): " + deltaY + " -> 적용값(mode): " + currentModeValue);
        }

        prevGesture = currentGesture;
    }

    private void applyDeviceMode(float finalModeValue) {
        if (actionListener != null) {
            actionListener.onDeviceModeApplied(finalModeValue);
        }
        System.out.println("LUMOS Core Engine -> 디바이스에 적용 완료: " + finalModeValue);
    }
}

File: app/src/main/java/io/github/adulescentia/LUMOS_lib/GestureStateManager.kt
package io.github.adulescentia.LUMOS_lib

class GestureStateManager {
    enum class Gesture {
        FIST,
        PALM,
        ONE_FINGER,
        V_SIGN,
        UNDEF
    }

    interface ActionListener {
        fun onDeviceSelectionToggled(isSelected: Boolean)
        fun onDevicePowerToggled()
        fun onDeviceModeApplied(modeValue: Float)
    }

    private var prevGesture = Gesture.UNDEF
    @JvmField
    val isDeviceSelected: Boolean = false
    @JvmField
    var isTrackingModeActive: Boolean = false
    private var readyForAction = true

    private val baseWristY = 0.0f
    private var currentModeValue = 0.0f
    private var sensitivity = 1.0f
    private var gestureQueue = SamplingQueue(3) { Gesture.UNDEF }

    private var actionListener: ActionListener? = null

    private var lastPerformedAction : Gesture = Gesture.UNDEF

    fun setActionListener(actionListener: ActionListener?) {
        this.actionListener = actionListener
    }

    fun setSensitivity(sensitivity: Float) {
        this.sensitivity = sensitivity
    }

    fun update(currentGesture: Gesture, currentWristY: Float) {
        if (isTrackingModeActive) {
            val deltaY = baseWristY - currentWristY
            currentModeValue = deltaY * sensitivity
            LumosLog.d(
                TAG,
                "🔄 [TRACKING] 실시간 높이 변화량 추적 중... deltaY(순수): $deltaY -> 적용값(mode): $currentModeValue"
            )
        }
        processGesture()
        gestureQueue.add(currentGesture)
    }
    private fun processGesture(){
        val reliableGesture = gestureQueue.checkDataIntegrity { it != Gesture.UNDEF && it != Gesture.FIST } ?: return
        val spamProofGesture = if (reliableGesture != lastPerformedAction) reliableGesture else Gesture.UNDEF
        if(spamProofGesture == Gesture.UNDEF) return
        lastPerformedAction = spamProofGesture
        when (spamProofGesture) {
            Gesture.ONE_FINGER -> {
                actionListener!!.onDeviceSelectionToggled(true)
            }
            Gesture.PALM -> {
                actionListener!!.onDevicePowerToggled()
            }
            Gesture.V_SIGN -> {
                if(isTrackingModeActive) applyDeviceMode(currentModeValue)
                isTrackingModeActive = !isTrackingModeActive
            }
            else -> {}
        }
    }

    private fun applyDeviceMode(finalModeValue: Float) {
        if (actionListener != null) {
            actionListener!!.onDeviceModeApplied(finalModeValue+1)
        }
        println("LUMOS Core Engine -> 디바이스에 적용 완료: ${finalModeValue+1}")
    }

    companion object {
        private const val TAG = "GestureStateManager"
    }
}
